tests/unit/classifier/svm/LibLinear_unittest.cc

```diff
@@ -16,1310 +15,456 @@
+		SGVector<float64_t> labels(num_samples);
+		for (index_t i = 0, j = 0; i < data.num_cols; ++i)
+		{
+			if (i % 2 == 0)
+				train_idx[j] = i;
+			else
+				test_idx[j++] = i;
 
-	ll->set_bias_enabled(false);
-	ll->set_features(train_feats);
-	ll->set_labels(ground_truth);
+			labels[i/2] = (i < data.num_cols/2) ? 1.0 : -1.0;
+		}
 
-	ll->set_liblinear_solver_type(liblinear_solver_type);
-	ll->train();
-	pred = ll->apply_binary(test_feats);
-	SG_REF(pred);
+		/*
+			Now fill in train_feats and test_feats, and ground_truth
+		*/
+		train_feats = (CDenseFeatures<float64_t>*)features.copy_subset(train_idx);
+		test_feats =  (CDenseFeatures<float64_t>*)features.copy_subset(test_idx);
+		ground_truth = new CBinaryLabels(labels);
 
-	liblin_accuracy = eval->evaluate(pred, ground_truth);
+		SG_REF(ground_truth);
+	}
+	void generate_data_l1()
+	{
+		initialize_train_and_test_and_ground();
 
-	EXPECT_NEAR(liblin_accuracy, 1.0, 1e-6);
 
-	SG_UNREF(ll);
-	SG_UNREF(train_feats);
-	SG_UNREF(test_feats);
-	SG_UNREF(ground_truth);
-	SG_UNREF(eval);
-	SG_UNREF(pred);
-}
+		transpose_feats_matrix();
 
-TEST(LibLinear,train_L2R_L2LOSS_SVC_DUAL)
-{
-	LIBLINEAR_SOLVER_TYPE liblinear_solver_type = L2R_L2LOSS_SVC_DUAL;
+	}
+	void generate_data_l2()
+	{
+		initialize_train_and_test_and_ground();
+	}
+	void initialize_train_and_test_and_ground_simple()
+	{
+		SGMatrix<float64_t> train_data(2, 10);
+		SGMatrix<float64_t> test_data(2, 10); //Always 2x10, doesn't matter l1 or l2
 
-	CDenseFeatures<float64_t>* train_feats = NULL;
-	CDenseFeatures<float64_t>* test_feats = NULL;
-	CBinaryLabels* ground_truth = NULL;
+		index_t num_samples = 10;
+		SGVector<int32_t> x{0,0,1,1,2,2,3,3,4,4,5,5,6,6,7,7,8,8,9,9};
+		SGVector<int32_t> y{0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1};
+		SGVector<int32_t> z{-1,1,-1,0,-1,-1,-2,0,0,-1,2,2,3,1,3,2,3,3,4,2};
 
-	generate_data_l2(train_feats, test_feats, ground_truth);
+		for(index_t i=0; i<num_samples*2; ++i){
+			train_data(y[i], x[i])=z[i];
+			test_data(y[i], x[i])=z[i];
+		}
 
-	CLibLinear* ll = new CLibLinear();
+		train_feats = new CDenseFeatures<float64_t>(train_data);
+		test_feats = new  CDenseFeatures<float64_t>(test_data);
 
-	CBinaryLabels* pred = NULL;
-	float64_t liblin_accuracy;
-	CContingencyTableEvaluation* eval = new CContingencyTableEvaluation();
+		SGVector<float64_t> labels(num_samples);
+		for (index_t i = 0; i < num_samples; ++i)
+		{
+			labels[i] = (i < num_samples/2) ? 1.0 : -1.0;
+		}
+		ground_truth = new CBinaryLabels(labels);
 
-	ll->set_bias_enabled(false);
-	ll->set_features(train_feats);
-	ll->set_labels(ground_truth);
+		SG_REF(train_feats);
+		SG_REF(test_feats);
+		SG_REF(ground_truth);
 
-	ll->set_liblinear_solver_type(liblinear_solver_type);
-	ll->train();
-	pred = ll->apply_binary(test_feats);
-	SG_REF(pred);
-	liblin_accuracy = eval->evaluate(pred, ground_truth);
+	}
+	void generate_data_l1_simple()
+	{
+			initialize_train_and_test_and_ground_simple();
```

Comment:
what about removing the _simple suffix?

Reply:
there is already initialize_train_and_test_and_ground() for the normal data, so I used the _simple suffix to indicate this is the simple data initializer (which is entirely different from the non simple one)

Reply:
Actually I agree, I got rid of both initialize_train_and_test_and_ground and initialize_train_and_test_and_ground_simple. Thanks!